ai_ta_backend/vector_database.py

```diff
@@ -982,17 +1069,29 @@ def getTopContexts(self, search_query: str, course_name: str, token_limit: int =
       start_time_overall = time.monotonic()
 
       found_docs: list[Document] = self.vector_search(search_query=search_query, course_name=course_name)
+      
+      # call context padding function here
+      final_docs = self.context_padding(found_docs, search_query, course_name)
+
+      print("FINAL DOCS: ", final_docs)
 
       pre_prompt = "Please answer the following question. Use the context below, called your documents, only if it's helpful and don't use parts that are very irrelevant. It's good to quote from your documents directly, when you do always use Markdown footnotes for citations. Use react-markdown superscript to number the sources at the end of sentences (1, 2, 3...) and use react-markdown Footnotes to list the full document names for each number. Use ReactMarkdown aka 'react-markdown' formatting for super script citations, use semi-formal style. Feel free to say you don't know. \nHere's a few passages of the high quality documents:\n"
       # count tokens at start and end, then also count each context.
       token_counter, _ = count_tokens_and_cost(pre_prompt + '\n\nNow please respond to my query: ' + search_query) # type: ignore
 
       valid_docs = []
       num_tokens = 0
-      for doc in found_docs:
-        doc_string = f"Document: {doc.metadata['readable_filename']}{', page: ' + str(doc.metadata['pagenumber']) if doc.metadata['pagenumber'] else ''}\n{str(doc.page_content)}\n"
+      
+      for doc in final_docs:
+        print("\n DOC: ", type(doc))
+        if isinstance(doc, dict):

```

Comment:
Is this check necessary? What's dictionary flow vs langchain flow?

Why do you have to modify the getTopContexts() function. Is it possible or prefereable to implement your function separately?

Reply:
There are 2 types of objects returned after context padding. The contexts retrieved from QDRANT vector search are of datatype `<class 'langchain.schema.document.Document'>` whereas the ones from Supabase are `<dict>`.  

The logic for context padding is - if chunk index exists, retrieve from Supabase. If chunk index does not exist (older documents), return the result from QDRANT search as it is. For QDRANT search result, the attributes are `page_content` and `metadata` which in turn has `readable_filename`, `url`, etc. Supabase results have a different structure - `context` containing `text`, `embeddings`, `page_number`, etc. I modified `getTopContexts()` to accomodate both of them.

Reply:
This is basically to handle contexts/documents which do not have `chunk_index`. Can you think of a better way to handle this? I was thinking since we have multi-query retrieval along with this, should we use parent-doc retrieval only for the newer documents (those with chunk indices)?